src/main/java/org/opensearch/ad/task/ADTaskManager.java

```diff
@@ -1132,52 +1200,35 @@ private void stopHistoricalAnalysis(
             );
     }
 
-    private boolean lastUpdateTimeExpired(ADTask adTask) {
+    private boolean lastUpdateTimeOfHistoricalTaskExpired(ADTask adTask) {
         // Wait at least 10 seconds. Piece interval seconds is dynamic setting, user could change it to a smaller value.
         int waitingTime = Math.max(2 * pieceIntervalSeconds, 10);
         return adTask.getLastUpdateTime().plus(waitingTime, ChronoUnit.SECONDS).isBefore(Instant.now());
     }
 
-    /**
-     * Check if AD task ended.
-     *
-     * @param adTask AD task
-     * @return true if task state is one of STOPPED, FINISHED or FAILED.
-     */
-    public boolean isADTaskEnded(ADTask adTask) {
-        return ADTaskState.STOPPED.name().equals(adTask.getState())
-            || ADTaskState.FINISHED.name().equals(adTask.getState())
-            || ADTaskState.FAILED.name().equals(adTask.getState());
-    }
-
-    private void resetTaskStateAsStopped(
+    private <T> void resetTaskStateAsStopped(
         ADTask adTask,
+        AnomalyDetectorFunction function,
         TransportService transportService,
-        boolean cleanDetectorCache,
-        AnomalyDetectorFunction function
+        ActionListener<T> listener
     ) {
-        String taskId = adTask.getTaskId();
-        adTask.setState(ADTaskState.STOPPED.name());
-        if (cleanDetectorCache) {
-            cleanDetectorCache(adTask, transportService, () -> resetTaskStateAsStopped(adTask, function, taskId));
-        } else {
-            resetTaskStateAsStopped(adTask, function, taskId);
-        }
-    }
-
-    private void resetTaskStateAsStopped(ADTask adTask, AnomalyDetectorFunction function, String taskId) {
-        Map<String, Object> updatedFields = new HashMap<>();
-        updatedFields.put(STATE_FIELD, ADTaskState.STOPPED.name());
-        updateADTask(adTask.getTaskId(), updatedFields, ActionListener.wrap(r -> {
-            logger.debug("Reset task state as STOPPED successfully for task {}", taskId);
-            if (function != null) {
-                function.execute();
-            }
-            if (ADTaskType.HISTORICAL_HC_DETECTOR.name().equals(adTask.getTaskType())) {
-                // Reset running entity tasks as STOPPED
-                resetEntityTasksAsStopped(adTask.getTaskId());
-            }
-        }, e -> logger.error("Failed to update task state as STOPPED for task " + taskId, e)));
+        cleanDetectorCache(adTask, transportService, () -> {
+            String taskId = adTask.getTaskId();
+            Map<String, Object> updatedFields = ImmutableMap.of(STATE_FIELD, ADTaskState.STOPPED.name());
+            updateADTask(taskId, updatedFields, ActionListener.wrap(r -> {
+                adTask.setState(ADTaskState.STOPPED.name());
+                if (function != null) {
+                    function.execute();
+                }
+                if (ADTaskType.HISTORICAL_HC_DETECTOR.name().equals(adTask.getTaskType())) {
```

Comment:
Why is only the historical HC task type considered here? What about RT HC entity tasks?

Reply:
For realtime anomaly detection, we only create detector level task, no entity level realtime task.

Reply:
could you add a comment for this?

Reply:
sure